File: backend/app/schemas/knowledge.py
"""Knowledge-base request/response schemas.

Every passage carries its source, its title and its URL. That is not decoration:
security advice a developer cannot trace back to who said it is advice they
cannot check, and this project's own notes must be visibly distinguishable from
MITRE's and OWASP's text.

``matched_by`` is exposed for the same reason. A passage found because it was
written for this exact rule and one found because it scored well against a
vector are both useful, and the reader is entitled to know which is which.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.knowledge import KnowledgeSource


class KnowledgePassage(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source: KnowledgeSource
    external_id: str
    document_title: str
    section: str
    text: str
    url: str | None
    source_version: str | None
    # Cosine similarity against the query, in [-1, 1]. Shown rounded; kept as a
    # float here so a test can assert an ordering rather than a rendering.
    score: float
    matched_by: str = Field(
        description="rule, cwe, owasp or semantic — why this passage was retrieved"
    )


class FindingKnowledgeResponse(BaseModel):
    finding_id: int
    rule_id: str
    cwe_id: str | None
    owasp_category: str | None
    # The text actually embedded to search with. Exposed because a retrieval
    # result nobody can reproduce is a result nobody can argue with.
    query: str
    passages: list[KnowledgePassage]


class KnowledgeStatusResponse(BaseModel):
    built: bool
    documents: int
    chunks: int
    embedded_chunks: int
    by_source: dict[str, int]
    embedding_models: list[str]
    source_versions: dict[str, str]
    built_at: datetime | None


__all__ = ["FindingKnowledgeResponse", "KnowledgePassage", "KnowledgeStatusResponse"]

File: frontend/src/types/knowledge.ts
/** Mirrors backend/app/schemas/knowledge.py */

export type KnowledgeSourceName = 'CWE' | 'OWASP' | 'SENTINELFORGE'

/** Why a passage was retrieved — shown so the reader can weigh it. */
export type MatchReason = 'rule' | 'cwe' | 'owasp' | 'semantic'

export interface KnowledgePassage {
  id: number
  source: KnowledgeSourceName
  external_id: string
  document_title: string
  section: string
  text: string
  url: string | null
  source_version: string | null
  score: number
  matched_by: MatchReason
}

export interface FindingKnowledgeResponse {
  finding_id: number
  rule_id: string
  cwe_id: string | null
  owasp_category: string | null
  /** The text that was embedded to search with, so a result can be reproduced. */
  query: string
  passages: KnowledgePassage[]
}

export interface KnowledgeStatus {
  built: boolean
  documents: number
  chunks: number
  embedded_chunks: number
  by_source: Record<string, number>
  embedding_models: string[]
  source_versions: Record<string, string>
  built_at: string | null
}

/** How each source is labelled in the UI.
 *
 * SentinelForge's own notes are named as ours. Presenting them under the same
 * badge as MITRE's text would borrow authority we have not earned.
 */
export const SOURCE_LABELS: Record<KnowledgeSourceName, string> = {
  CWE: 'MITRE CWE',
  OWASP: 'OWASP',
  SENTINELFORGE: 'SentinelForge note',
}

export const MATCH_LABELS: Record<MatchReason, string> = {
  rule: 'written for this rule',
  cwe: 'this weakness',
  owasp: 'this OWASP category',
  semantic: 'related',
}
